.github/workflows/ci.yml

```diff
@@ -0,0 +1,327 @@
+          VAR=${PREFIX}/var
+          mkdir -p $BIN
+          mkdir -p $VAR
+
+          echo "prefix=${PREFIX}" >> "$GITHUB_OUTPUT"
+          echo "bin=${BIN}" >> "$GITHUB_OUTPUT"
+          echo "bin=${BIN}" >> "$GITHUB_PATH"
+          echo "var=${VAR}" >> "$GITHUB_OUTPUT"
+          cat "$GITHUB_OUTPUT"
+
+      # Provides module, language version and testpath from build name
+      - name: Output build metadata
+        id: metadata
+        run: |
+          # Split the build name by '-' into module and version
+          IFS='-' read -r module version <<< "${{ matrix.build }}"
+
+          testpath="test/test_${module}*"
+
+          # Run all tests for "unit" and "python"
+          # Python is the default module for tests
+          if [ "$module" = "unit" ] || [ "$module" = "python" ]; then
+            testpath="test"
+          fi
+
+          echo "module=${module}" >> "$GITHUB_OUTPUT"
+          echo "version=${version}" >> "$GITHUB_OUTPUT"
+          echo "testpath=${testpath}" >> "$GITHUB_OUTPUT"
+
+          NJS_VERSION=$(sed -n "s/NJS_VERSION := \(.*\)/\1/p" pkg/contrib/src/njs/version)
+          echo "njs_version=${NJS_VERSION}" >> "$GITHUB_OUTPUT"
+
+          cat "$GITHUB_OUTPUT"
+
+      # https://github.com/actions/runner-images/issues/2821
+      - name: Kill mono process
+        run: |
+          sudo systemctl stop mono-xsp4.service
+          sudo systemctl mask mono-xsp4.service
+          sudo systemctl status mono-xsp4.service || true
+          PID=$(sudo lsof -t -i :8084)
+          echo "Killing PID $PID"
+          sudo kill -9 $PID
+
+      ##
+      ## njs
+      ##
+
+      - name: Clone njs repository
+        uses: actions/checkout@v4
+        with:
+          repository: nginx/njs
+          ref: '${{ steps.metadata.outputs.njs_version }}'
+          path: njs
+
+      - name: Make njs
+        run: |
+          ./configure --no-libxml2 --no-zlib
+          make -j4 -k
+        working-directory: njs
+
+      ##
+      ## Unit
+      ##
+
+      - name: Configure unit
+        run: |
+          ./configure \
+            --prefix=${{ steps.dir.outputs.prefix }} \
+            --sbindir=${{ steps.dir.outputs.bin }} \
+            --logdir=${{ steps.dir.outputs.var }}/log \
+            --log=${{ steps.dir.outputs.var }}/log/unit/unit.log \
+            --runstatedir=${{ steps.dir.outputs.var }}/run \
+            --pid=${{ steps.dir.outputs.var }}/run/unit/unit.pid \
+            --control=unix:${{ steps.dir.outputs.var }}/run/unit/control.sock \
+            --modules=${{ steps.dir.outputs.prefix }}/lib/unit/modules \
+            --statedir=${{ steps.dir.outputs.var }}/state/unit \
+            --tests \
+            --openssl \
+            --njs \
+            --cc-opt="-I njs/src/ -I njs/build"  \
+            --ld-opt="-L njs/build" \
+            --debug
+
+      - name: Make unit
+        run: |
+          make -j4 -k
+
+      ##
+      ## Go
+      ##
+
+      - uses: actions/setup-go@v4
```

Comment:
It looks like this installs some version of Go that is not shipped with Ubuntu 22.04 but somewhere else. (I think it also applies to other language modules, except for Perl).

On Buildbot we use whatever is shipped on that OS (with a notable exception of Node, which we fetch from nodesource repos).  This helps us ensure our software is compatible with what is broadly used by respective OS users.

Is the intention here to test with with whatever is a "current version" of a language available from actions vendors?  If so, when a new incompatible version of a language is published it means our pipelines will be red and prevent (at least formally) from unrelated PRs to be merged?  I can see some value in having runs with new versions though.

Should we stick to specific known good versions instead?  Should we not be using those "edge" versions but revert to what is available in the OS packages repositories?

Reply:
Oh, I've just read the commit message that describes the versioning, so please disregard that part of the question.  The OS-provided vs actions-provided still stands, though.

Reply:
I'm on holiday and setting a bad example, but my preference for the testing aligns with Dylan's:

- I want us to easily test against arbitrary versions of upstream runtimes, separate from distro-provided versions.

This isn't exclusive to testing against distro-provided dependencies; we should _also_ be testing that we successfully build on platforms we want to support, but that can be a second phase of this effort. For now, the existing BuildBot infrastructure covers us there.

Reply:
Thanks for clarifications @arbourd!  I am not against these changes, but I wanted to have clarity on what this PR tries to achieve.

> > Should we stick to specific known good versions instead? Should we not be using those "edge" versions but revert to what is available in the OS packages repositories?
> 
> Is there any reason not to do both? Is it an XOR? I hope I've laid out a competent argument about why I think testing against the non-default versions is important for Unit. Is there any context I'm missing that makes this untenable?

Nope, I think this is a correct approach.  We can start with the tests that make sense from development point of view (new or "currently stable" language versions), and then build up other parts of CI that will test the changes on what's shipped in the OS we need to support, just like we do on BB.

To add some info, for BB, we kinda sorta got away with testing our code on platforms like Alpine and FreeBSD that typically offer up-to-date language runtimes.

From my point of view, this conversation can be marked as resolved.